### ISSUE TITLE
Support Vulkan Video for Hardware Acceleration

### ISSUE THREAD

**Issue by gitpubber** [mpv-player/mpv#9675]
Vulkan Video extensions allow for GPU-accelerated video encode/decode. It can be considered alternative to VAAPI  and Nvencode/decode . It's still early stages for the API but i would like "Vulkan Video"  to be implemented in mpv. Latest Nvidia drivers do support this API and Intel/AMD might implement this API in future updates of the drivers. Gstreamer already is exploring this new API.

https://www.khronos.org/blog/an-introduction-to-vulkan-video

https://www.phoronix.com/scan.php?page=news_item&px=Vulkan-Video-Provisional-Spec

https://blogs.igalia.com/vjaquez/2021/07/09/video-decoding-in-gstreamer-with-vulkan/

https://github.com/nvpro-samples/vk_video_samples

https://www.phoronix.com/scan.php?page=news_item&px=Vulkan-Video-XDC2021

**Comment by jeeb** [comment]
This would go through FFmpeg HWAccels and there is some [WIP work](https://github.com/cyanreg/FFmpeg/commits/vulkan_decode) ongoing with that. Do note that the main gain is that you (at least are supposed to) get a Vulkan surface from the hwdec API, instead of something you'd have to import (like VAAPI DRM or nvdec CUDA surface).

And of course the possibility that you'd get all of Intel, AMD and nvidia from a single API... if they all implement it widely enough :P

Also would be a nice surprise if nvidia finally released stable drivers with this feature, that has so far been the main reason for me not testing this out as I only have a single workstation with nvidia and I'd rather not utilize development drivers on that.

**Comment by lextra2** [comment]
I wouldn't hold my breath on amd doing anything for video encoding/decoding on gpus. They do the bare minimum as is.

And I don't see nvidia contribute to this new vulkan thing either. And rightfully so, because anything cuda runs circles around the few alternatives.

**Comment by philipl** [comment]
nvidia has the only working implementation of it.

**Comment by lextra2** [comment]
> nvidia has the only working implementation of it.

Good to know.

**Comment by Pentaphon** [comment]
> I wouldn't hold my breath on amd doing anything for video encoding/decoding on gpus. They do the bare minimum as is.

They have to do something with Vulkan Video if they want to remain competitive. It's only a matter of time.

**Comment by berniyh** [comment]
> I wouldn't hold my breath on amd doing anything for video encoding/decoding on gpus. They do the bare minimum as is.

There you go, it was just merged into mesa and will be part of the next version (mesa-23.1). ;)
https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/20388

Of course, it's early stages and not actually by an AMD person, but it's the most relevant Vulkan driver for AMD cards right now. Also only decoding for now.

**Comment by Jules-A** [comment]
They were probably waiting on VP9/AV1 support before doing anything. With just H264/H265 it's honestly not too useful, would MPV even implement it with just them or would they wait for VP9/AV1?

**Comment by CoelacanthusHex** [comment]
> They were probably waiting on VP9/AV1 support before doing anything.

If an unsupported format is used, MPV will skip this backend and then try the next one. Just like we do in other backends.

> With just H264/H265 it's honestly not too useful, would MPV even implement it with just them or would they wait for VP9/AV1?

In fact, H264 and H265 are the encodings I use most often with MPV. Most animations, movies, and TV dramas are encoded with these two codecs. On Bilibili, the most commonly used video website in China mainland, most videos only support H264 and H265. I basically use VP9 and AV1 only when watching Youtube.

**Comment by Jules-A** [comment]
> On Bilibili, the most commonly used video website in China mainland, most videos only support H264 and H265.

Weird, whenever I watch something on Bilibili it's always AV1 in the last year or so even though I only have an RX570. I don't have the paid version of the site though so I have no idea what it would serve at 4k (since 4k is too much for my CPU to decode in AV1).

> If an unsupported format is used, MPV will skip this backend and then try the next one. Just like we do in other backends.

Thanks, I wasn't aware but how does it know what backend to try next and is there even an option to set fallbacks?

**Comment by CoelacanthusHex** [comment]
> Thanks, I wasn't aware but how does it know what backend to try next and is there even an option to set fallbacks?

It has a builtin list to try. It will try all possible decoder but it seems no a option to set fallback order. I think the default list fit the most situation. If you still think it's necessary to be able to control the order, you can open a feature request.

> Weird, whenever I watch something on Bilibili it's always AV1 in the last year or so even though I only have an RX570. I don't have the paid version of the site though so I have no idea what it would serve at 4k (since 4k is too much for my CPU to decode in AV1).

It's related to the uploader and upload time. It only apply to recently (after 2022) video, and only apply to the most popular up at beginning. So most videos on bilibili still H264, especially old Anime and movie.

**Comment by 24fpsDaVinci** [comment]
> only NVIDIA offers official support for it at the moment

gotcha, this is needed for intel users 
https://airlied.blogspot.com/2022/12/vulkan-video-decoding-anv-update.html
